libp2p/multiaddress.nim

```diff
@@ -667,200 +689,239 @@ proc validate*(ma: MultiAddress): bool =
-      copyMem(addr data[0], unsafeAddr value[0], len(value))
-      result.data.writeSeq(data)
-    result.data.finish()
+    err("Protocol not found")
+  else:
+    var res: MultiAddress
+    res.data = initVBuffer()
+    res.data.writeVarint(cast[uint64](proto.mcodec))
+    if proto.kind in {Fixed, Length, Path}:
+      if len(value) == 0:
+        err("Value must not be empty array")
+      else:
+        if proto.kind == Fixed:
+          res.data.writeArray(value)
+        else:
+          var data = newSeq[byte](len(value))
+          copyMem(addr data[0], unsafeAddr value[0], len(value))
+          res.data.writeSeq(data)
+        res.data.finish()
+        ok(res)
 
-proc init*(mtype: typedesc[MultiAddress], protocol: MultiCodec,
-           value: PeerID): MultiAddress {.inline.} =
+proc init*(mtype: typedesc[MultiAddress], protocol: MultiCodec, value: PeerID): MAResult[MultiAddress] {.inline.} =
   ## Initialize MultiAddress object from protocol id ``protocol`` and peer id
   ## ``value``.
   init(mtype, protocol, value.data)
 
-proc init*(mtype: typedesc[MultiAddress], protocol: MultiCodec): MultiAddress =
+proc init*(mtype: typedesc[MultiAddress], protocol: MultiCodec): MAResult[MultiAddress] =
   ## Initialize MultiAddress object from protocol id ``protocol``.
   let proto = CodeAddresses.getOrDefault(protocol)
   if proto.kind == None:
-    raise newException(MultiAddressError, "Protocol not found")
-  result.data = initVBuffer()
-  if proto.kind != Marker:
-    raise newException(MultiAddressError, "Protocol missing value")
-  result.data.writeVarint(cast[uint64](proto.mcodec))
-  result.data.finish()
+    err("Protocol not found")
+  else:
+    var res: MultiAddress
+    res.data = initVBuffer()
+    if proto.kind != Marker:
+      raise newException(MultiAddressError, "Protocol missing value")
+    res.data.writeVarint(cast[uint64](proto.mcodec))
+    res.data.finish()
+    ok(res)
 
-proc init*(mtype: typedesc[MultiAddress], protocol: MultiCodec,
-           value: int): MultiAddress =
+proc init*(mtype: typedesc[MultiAddress], protocol: MultiCodec, value: int): MAResult[MultiAddress] =
   ## Initialize MultiAddress object from protocol id ``protocol`` and integer
   ## ``value``. This procedure can be used to instantiate ``tcp``, ``udp``,
   ## ``dccp`` and ``sctp`` MultiAddresses.
   var allowed = [multiCodec("tcp"), multiCodec("udp"), multiCodec("dccp"),
                  multiCodec("sctp")]
   if protocol notin allowed:
-    raise newException(MultiAddressError,
-                       "Incorrect protocol for integer value")
-  let proto = CodeAddresses.getOrDefault(protocol)
-  result.data = initVBuffer()
-  result.data.writeVarint(cast[uint64](proto.mcodec))
-  if value < 0 or value > 65535:
-    raise newException(MultiAddressError, "Incorrect integer value")
-  result.data.writeArray(toBytesBE(cast[uint16](value)))
-  result.data.finish()
+    err("Incorrect protocol for integer value")
```

Comment:
elsewhere, we've prefixed the error string with a short module name, in case the string is displayed anywhere - this seemed like a pragmatic approach